dir.c

```diff
@@ -3626,6 +3651,8 @@ rb_dir_s_empty_p(VALUE obj, VALUE dirname)
 void
 Init_Dir(void)
 {
+    rb_gc_register_address(&chdir_lock.path);
```

Comment:
```suggestion
    rb_gc_register_address(&chdir_lock.path);
    rb_gc_register_address(&chdir_lock.thread);
```

Without this, if GC compaction moves the thread object, comparisons with the current thread would fail. I guess it's also a problem currently but no one is compacting inside chdir 😅